File: Loenn/entities/Distort.lua
local drawableSpriteStruct=require("structs.drawable_sprite")
local Distort={}
Distort.name="DBBHelper/Distort"

Distort.fieldOrder={
    "x","y","path","scale","rotation","waveStrength","distortStrength","DrawMask","DebugMask"
}
Distort.fieldInformation={

    scale={
        minimumValue=0.02
    },
    distortStrength={
        minimumValue=0.0,
        maximumValue=1.0
    },
    waveStrength={
        minimumValue=0.0,
        maximumValue=1.0
    }

}
Distort.placements={
    name="Distort",
    data={
        path="objects/DBB_Items/Distort/sphere_gradient",
        scale=1.0,
        rotation=0.0,
        distortStrength=0.0,
        waveStrength=1.0,
        DrawMask=false,
        DebugMask=false
    }
}

function Distort.sprite(room, entity)
    local texture1=entity.path
    local texture2="objects/DBB_Items/DBBHelperLogo/ProfileWhiteLogo"

    local sprite1=drawableSpriteStruct.fromTexture(texture1,entity)
    local sprite2=drawableSpriteStruct.fromTexture(texture2,entity)

    sprite1.rotation=-entity.rotation/180.0*math.pi
    sprite2.rotation=-entity.rotation/180.0*math.pi

    sprite1:setScale(entity.scale,entity.scale)
    sprite2:setScale(0.05,0.05)
    return {sprite1,sprite2}
end
return Distort

File: Source/DBB_VanillaEntity/Distort.cs
using System;
using Celeste.Mod.Entities;
using Microsoft.Xna.Framework;
using Monocle;
namespace Celeste.Mod.DBBHelper.Entities {

    [CustomEntity("DBBHelper/Distort")]
    public class Distort:Entity
    {
        //------------一些通用参数------------
        private float distort_strength=1.0f;//扭曲效果的力度
        private float wave_strength=1.0f;//波动效果的力度
        private float scale=1.0f;//图片缩放比例
        private float rotation=0.0f;//图片旋转角度
        private bool draw_mask=false;//实际绘制时是否应当把掩码图片也绘制出来
        private bool draw_mask_when_debug=false;//Debug时是否应当把掩码图片绘制出来
        //------------一些内部参数------------
        private string path="";//掩码图片的路径
        private MTexture distort_texture=null;//贴图
        public Distort(EntityData data,Vector2 offset)
        {
            Position=data.Position+offset;

            path = data.Attr("path");
            distort_strength = data.Float("distortStrength");
            wave_strength = data.Float("waveStrength");
            scale = data.Float("scale");
            rotation = -data.Float("rotation") / 180.0f * (float)Math.PI;
            draw_mask = data.Bool("DrawMask");
            draw_mask_when_debug = data.Bool("DebugMask");
            distort_texture = GFX.Game[path];
            if(GFX.Game.Has(path)==false)
            {
                distort_texture=GFX.Game["objects/DBB_Items/Distort/sphere_gradient"];
            }
            Depth=-100;
        }
        public override void Added(Scene scene)
        {
            base.Added(scene);
            if(distort_texture!=null)
            {
                Add(new DisplacementRenderHook(RenderDisplacement));
            }
        }
        public void RenderDisplacement()
        {
            distort_texture.DrawCentered(Position,new Color(wave_strength,wave_strength,wave_strength,distort_strength),scale,rotation);   
        }
        public override void DebugRender(Camera camera)
        {
            base.DebugRender(camera);
            if(draw_mask_when_debug==true)
            {
                distort_texture.DrawOutlineCentered(Position,new Color(1.0f,1.0f,1.0f,1.0f),scale,rotation);  
            }
        }
        public override void Render()
        {
            base.Render();
            if(draw_mask==true)
            {
                distort_texture.DrawCentered(Position,new Color(1.0f,1.0f,1.0f,1.0f),scale,rotation);
            }
        }
    }


}